File: apps/api/src/middleware/auth.middleware.ts
import type { NextFunction, Request, Response } from "express";

import { AppError } from "../utils/app-error.js";
import { authenticateAccessToken } from "../services/session-auth.service.js";

declare module "express-serve-static-core" {
  interface Request {
    auth?: {
      userId: string;
      sessionId: string;
    };
  }
}

export async function requireAuth(request: Request, _response: Response, next: NextFunction) {
  const token = request.headers.authorization?.replace("Bearer ", "");

  if (!token) {
    next(new AppError("Authentication required", 401));
    return;
  }

  try {
    request.auth = await authenticateAccessToken(token);
    next();
  } catch {
    next(new AppError("Invalid or expired token", 401));
  }
}

export async function optionalAuth(request: Request, _response: Response, next: NextFunction) {
  const token = request.headers.authorization?.replace("Bearer ", "");

  if (!token) {
    next();
    return;
  }

  try {
    request.auth = await authenticateAccessToken(token);
    next();
  } catch {
    next(new AppError("Invalid or expired token", 401));
  }
}

File: apps/api-go/internal/transport/http/auth_middleware.go
package httptransport

import (
	"net/http"
	"strings"

	authapp "github.com/mackings/histora/apps/api-go/internal/app/auth"
	"github.com/mackings/histora/apps/api-go/internal/shared/appctx"
	"github.com/mackings/histora/apps/api-go/internal/shared/apperror"
	"github.com/mackings/histora/apps/api-go/internal/shared/response"
)

func RequireAuth(authService *authapp.Service) func(http.Handler) http.Handler {
	return func(next http.Handler) http.Handler {
		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			tokenValue := bearerToken(r)
			if tokenValue == "" {
				response.Error(w, apperror.Unauthorized("Authentication required"))
				return
			}
			userID, sessionID, err := authService.VerifyAccessToken(tokenValue)
			if err != nil {
				response.Error(w, apperror.Unauthorized("Invalid or expired access token"))
				return
			}
			next.ServeHTTP(w, r.WithContext(appctx.WithAuthUser(r.Context(), appctx.AuthUser{ID: userID, SessionID: sessionID})))
		})
	}
}

func OptionalAuth(authService *authapp.Service) func(http.Handler) http.Handler {
	return func(next http.Handler) http.Handler {
		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			tokenValue := bearerToken(r)
			if tokenValue == "" || authService == nil {
				next.ServeHTTP(w, r)
				return
			}
			userID, sessionID, err := authService.VerifyAccessToken(tokenValue)
			if err != nil {
				next.ServeHTTP(w, r)
				return
			}
			next.ServeHTTP(w, r.WithContext(appctx.WithAuthUser(r.Context(), appctx.AuthUser{ID: userID, SessionID: sessionID})))
		})
	}
}

func bearerToken(r *http.Request) string {
	header := r.Header.Get("Authorization")
	if header == "" {
		return ""
	}
	parts := strings.SplitN(header, " ", 2)
	if len(parts) != 2 || !strings.EqualFold(parts[0], "Bearer") {
		return ""
	}
	return strings.TrimSpace(parts[1])
}
